### ISSUE TITLE
Thinking of packet processing logic.

### ISSUE THREAD

**Issue by TimonPost** [amethyst/laminar#39]
I am thinking of the packet workflow under which you can think: easiest way for creating packets and sending them over to the other side with all the options that a packet could have like encryption etc.

Take the following things into consideration:
1. Options like encryption/decryption
2. Options for different reliability strategies: Unreliable, ReliableUnordered, Sequenced ReliableOrdered.
3. Must be configurable.
4. Must be easily extendable.
5. The user does not have to be bothered with all the different network tactics and implementations on how to send packets. The only thing a user should care about is, sending bytes!

Here I have a draft on how to separate packets of different types. And on how we can convert a simple user packet to some more complex packet types about which the user should not care about.

https://gist.github.com/TimonPost/303c53fa1454d724120a1173b04a9745

Now working on processing the bytes of each packet this issue will be updated soon.

**Comment by OvermindDL1** [comment]
In addition to priorities to determine which to focus on sending first, and also channels so you can have multiple 'ordered streams' in progress at the same time.

**Comment by TimonPost** [comment]
There was once some kind of implementation of channels. But some folks did not like opening different UPD sockets for different kind of implementations. See [this](https://github.com/LucioFranco/amethyst-protocol/pull/21) PR

**Comment by OvermindDL1** [comment]
> But some folks did not like opening different UPD sockets for different kind of implementations.

Because you shouldn't need to.  It should be part of the application packet header, using a single byte is well worth channel support for 255 different channels for anything sequenced or ordered (and unordered doesn't need it at all).

**Comment by TimonPost** [comment]
I like the channel idea but am not sure if I understand it right. Can you be more specific with the channels a bit? 

Currently, how I am imagine this is having channels who will process each packet on its own way. Based on the id of the packet it will be processed by a certain channel. Then my next questions are:

1. How is the user able to get the packet from the channel. 
2. Does the user have access to the channel contents? And processing stuff.
3. How does the user get the newest packets from the channels? 
4. Do you have some example of the channels Raknet is using? I have been wandering around in their code but not have seen anything related to channels.

**Comment by OvermindDL1** [comment]
In the Raknet docs it speaks of the costs (my bolding):  http://www.raknet.net/raknet/manual/systemoverview.html

* Per datagram:
  1. 1 byte for bitflags
  2. 4 bytes for timestamp, used to calculate RTT for congestion control. RakNet version 4 does not transmit this field.
  3. 3 bytes for sequence number, used to lookup datagrams for ACKs

* Per message:
  1. 1 byte for bitflags
  2. 2 bytes for message length
  3. if (RELIABLE,RELIABLE_SEQUENCED,RELIABLE_ORDERED as specified in bitflags)
    A. 3 bytes for sequence number, used to prevent returning to the user duplicated messages
  4. if (UNRELIABLE_SEQUENCED,RELIABLE_SEQUENCED,RELIABLE_ORDERED as specified in bitflags)
    A. 3 bytes for sequence number, used to reorganize messages on the same channel in order
    B. **1 byte for ordering channel**
  5. if (message over MTU as specified in bitflags)
    A. 4 bytes for number of splits, not compressed to improve performance
    B. 2 bytes for identifier for which split this is
    C. 4 bytes for index into number of splits, not compressed to improve performance

**Comment by OvermindDL1** [comment]
At http://www.raknet.net/raknet/manual/detailedimplementation.html you can see in the `Sending Data` section:

The best way to illustrate sending data is with an example:
const char* message = "Hello World";

To all connected systems:
peer->Send((char*)message, strlen(message)+1, HIGH_PRIORITY, RELIABLE, 0, UNASSIGNED_RAKNET_GUID, true);

The first parameter is your data and must be a byte stream. Since we're sending a string, and a string is a byte stream, we can send it directly without any casting.

The second parameter is how many bytes to send. In this example we send the length of the string and one more for the null terminator.

The third parameter is the priority of the packet. This takes one of three values:
IMMEDIATE_PRIORITY,
HIGH_PRIORITY
MEDIUM_PRIORITY
LOW_PRIORITY

IMMEDIATE_PRIORITY messages signal RakNet's update thread to update immediately. Assuming bandwidth is available, they get sent immediately. HIGH_PRIORITY, MEDIUM_PRIORITY, and LOW_PRIORITY messages are put into a buffer. The next time RakNet's update thread ticks (which is every 10 milliseconds) those messages get sent. Those 3 priorities can be more efficient for bandwidth, because if multiple messages can be aggregated into a single datagram, RakNet will transparently do so

Twice as many messages get sent out for each higher priority. So if messages of all priorities were waiting to go out, 8 IMMEDIATE_PRIORITY, 4 HIGH_PRIORITY, 2 MEDIUM_PRIORITY, and 1 LOW_PRIORITY would be sent. Obviously however, if only LOW_PRIORITY messages were waiting at all, then those messages would all go out as fast as possible.

The fourth parameter takes one of five major values. Lets say you send data 1,2,3,4,5,6. Here's the order and substance of what you might get back:

UNRELIABLE - 5, 1, 6
UNRELIABLE_SEQUENCED - 5 (6 was lost in transit, 1,2,3,4 arrived later than 5)
RELIABLE - 5, 1, 4, 6, 2, 3
RELIABLE_ORDERED - 1, 2, 3, 4, 5, 6
RELIABLE_SEQUENCED - 5, 6 (1,2,3,4 arrived later than 5)

For more details on this refer to PacketPriority.h.

**The fifth parameter to Send() (0 in this example) is which ordering stream to use. This is used for relative ordering of packets in relation to other packets on the same stream. It's not important for now, but for more information on this refer to the Sending Packets section.**

The sixth parameter (UNASSIGNED_RAKNET_GUID), is the remote system to send to. UNASSIGNED_RAKNET_GUID is a reserved value meaning "no-one in particular". This parameter means one of two things : either who you want to send the packet to, or who you don't want to send the packet to, depending on the value of broadcast, which is the last parameter.

The seventh parameter (true in this example) is whether to broadcast to all connected systems or not. This parameter works with the sixth parameter. If broadcast is true, then the sixth parameter specifies who not to send to. If it is false, then it specifies who to send to. If we want to broadcast to everyone, then we just specify UNASSIGNED_RAKNET_GUID. This works out when relaying packets, because the Packet::systemAddress field in the packet you get will specify who the sender is. We can relay the packet to everyone BUT the sender, which makes sense since we usually don't want to send the same information back to the person who just sent it to us.

**Comment by OvermindDL1** [comment]
And you can see how `send` is defined at:  http://www.jenkinssoftware.com/raknet/manual/Doxygen/classRakNet_1_1RakPeerInterface.html#6d9a5415556a9c138854cc05c707b8e7

```cpp
uint32_t RakNet::RakPeerInterface::Send(
    const RakNet::BitStream * | bitStream,
    PacketPriority | priority,
    PacketReliability | reliability,
    char | orderingChannel,
    const AddressOrGUID | systemIdentifier,
    bool | broadcast,
    uint32_t | forceReceiptNumber = 0
)
```

With the documentation of:

In/Out | Type | Description
-- | -- | --
[in] | bitStream | The bitstream to send
[in] | priority | What priority level to send on. See PacketPriority.h
[in] | reliability | How reliability to send this data. See PacketPriority.h
[in] | orderingChannel | When using ordered or sequenced messages, what channel to order these on. Messages are only ordered relative to other messages on the same stream
[in] | systemIdentifier | Who to send this packet to, or in the case of broadcasting who not to send it to. Pass either a SystemAddress structure or a RakNetGUID structure. Use UNASSIGNED_SYSTEM_ADDRESS or to specify none
[in] | broadcast | True to send this packet to all connected systems. If true, then systemAddress specifies who not to send the packet to.
[in] | forceReceipt | If 0, will automatically determine the receipt number to return. If non-zero, will return what you give it.

**Comment by TimonPost** [comment]
Thanks for the detailed reaction, forgot to react on this issue. I am currently investigating RakNet. O man it is so much code :). But I have to take note here the most of RakNet its features are going way beyond or usecases. The only thing we need is the reliability implementation.

The file containing the most concerns for us can be found [here](https://github.com/SLikeSoft/SLikeNet/blob/master/Source/src/ReliabilityLayer.cpp).

**Comment by OvermindDL1** [comment]
> Thanks for the detailed reaction, forgot to react on this issue. I am currently investigating RakNet. O man it is so much code :). But I have to take note here the most of RakNet its features are going way beyond or usecases. The only thing we need is the reliability implementation.

Correct, RakNet is a *LOT* of things built on a comparatively simple (okay not really that simple, but still) network interface, should just focus on that to start with.  I can still pick up the code base pretty quickly if you have questions about anything in it though.

**Comment by TimonPost** [comment]
This kind of is solved #65
